### PR TITLE
Propagate TX context in Kotlin coroutines

### LINKED ISSUE #1216: Breaking Change for Kotlin coroutines - "Async return type doesn't support transactional execution."
### Issue description

The [Micronaut Data 3.1.0 release](https://github.com/micronaut-projects/micronaut-data/releases/tag/v3.1.0) added Kotlin coroutine support.

This introduced a breaking change. Previously, one could use `suspend` functions in Micronaut Data `@Repository` interfaces just fine, though now attempting to do so can result in a `ConfigurationException` stating that "Async return type doesn't support transactional execution."

See the discussion [here](https://github.com/micronaut-projects/micronaut-data/pull/1165/files#r746080283) for more details, but in short, the prior suspending functions weren't _actually_ async but were instead using a `SynchronousTransactionManager`, and that "support" was removed.

Full disclosure, I'm doing this writeup expecting no code to be changed (though it'd be nice if this was documented as a breaking change or part of the documented requirements). I'm writing this so that Google searches will point developers to why their upgrade to 3.1.0+ isn't working.

All that said, hypothetical other developers, here are two workarounds:

1) Use a reactive transaction manager like [R2DBC](https://micronaut-projects.github.io/micronaut-r2dbc/latest/guide/). I haven't used this yet myself, but TransactionalInterceptor demands a [`ReactiveTransactionOperations` implementation](https://github.com/micronaut-projects/micronaut-data/pull/1165/files#diff-ad46901a48c0483178545c356c5828a4ebd2e0c5aa3f5b4c27cb4858d6973e77R104) to avoid throwing the `ConfigurationException`, and [DefaultR2dbcRepositoryOperations](https://github.com/micronaut-projects/micronaut-data/blob/8649d853b2a9096e259b234aef471c79da0a04c1/data-r2dbc/src/main/java/io/micronaut/data/r2dbc/operations/DefaultR2dbcRepositoryOperations.java#L130) is the only class that implements that interface so far.
2) Convert all your suspending functions in your repositories to non-suspending.

### LINKED ISSUE #1271: @Transactional not working when using Coroutines
### Expected Behavior

`@Transactional` should work. For example, no database rows should be saved when an exception is thrown in a method wrapped with `@Transactional`.

### Actual Behaviour

When using `@Transactional` on a suspending function, database rows are still being saved even when an exception is thrown.

### Steps To Reproduce

I've reproduced the issue in this repo, and included instructions: https://github.com/joshchoo/micronaut-coroutines-transactional-issue

The example application saves a database entity **twice** in a `@Transactional`, which throws an duplicate key value exception.

### Environment Information

- Operating System: MacOS 11.6.2
- JDK details:
```
openjdk version "17.0.1" 2021-10-19
OpenJDK Runtime Environment Temurin-17.0.1+12 (build 17.0.1+12)
OpenJDK 64-Bit Server VM Temurin-17.0.1+12 (build 17.0.1+12, mixed mode, sharing)
```

### Example Application

https://github.com/joshchoo/micronaut-coroutines-transactional-issue

### Version

3.2.6

### PR DESCRIPTION
**Pull request by dstepanov**
Fixes #1271
Fixes #1216

### CODE REVIEW

**Comment by graemerocher** [comment]
And what about R2DBC and coroutines?

File: data-tx/src/main/java/io/micronaut/transaction/interceptor/CoroutineTxHelper.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * Helper to setup Kotlin coroutine context.
+ *
+ * @author Denis Stepanov
+ * @since 3.3
+ */
+@Internal
+@Singleton
+@Requires(classes = kotlin.coroutines.CoroutineContext.class)
+public final class CoroutineTxHelper {
```
Comment (graemerocher) [inline data-tx/src/main/java/io/micronaut/transaction/interceptor/CoroutineTxHelper.java:19]: Does this need to be public?

File: data-tx/src/main/java/io/micronaut/transaction/interceptor/TransactionalInterceptor.java [hunk 2]
```diff
@@ -79,14 +78,26 @@ public String toString() {
 
     @NonNull
     private final BeanLocator beanLocator;
+    private final Optional<CoroutineTxHelper> coroutineTxHelper;
```
Comment (graemerocher) [inline data-tx/src/main/java/io/micronaut/transaction/interceptor/TransactionalInterceptor.java:81]: Don't store optional fields, just store either the value or null

**Review by graemerocher — CHANGES_REQUESTED** [review]

**Comment by dstepanov** [comment]
R2dbc TX propagation is a bit complicated, I will try to add it in the future

**Review by graemerocher — APPROVED** [review]

**Pull request merged**